db/crud.go

```diff
@@ -633,37 +633,13 @@ func (db *Database) OnDemandImportForWrite(docid string, doc *Document, deleted
 	return nil
 }
 
-// Updates or creates a document.
-// The new body's BodyRev property must match the current revision's, if any.
-func (db *Database) Put(docid string, body Body) (newRevID string, doc *Document, err error) {
-	// Get the revision ID to match, and the new generation number:
-	matchRev, _ := body[BodyRev].(string)
-	generation, _ := ParseRevID(matchRev)
-	if generation < 0 {
-		return "", nil, base.HTTPErrorf(http.StatusBadRequest, "Invalid revision ID")
-	}
-	generation++
-	deleted, _ := body[BodyDeleted].(bool)
-
-	expiry, err := body.ExtractExpiry()
-	if err != nil {
-		return "", nil, base.HTTPErrorf(http.StatusBadRequest, "Invalid expiry: %v", err)
-	}
+func (db *Database) Put(newDoc *Document) (newRevID string, doc *Document, err error) {
 
-	// Create newDoc which will be used to pass around Body
-	newDoc := &Document{
-		ID: docid,
-	}
-
-	// Pull out attachments
-	newDoc.DocAttachments = GetBodyAttachments(body)
-	delete(body, BodyAttachments)
-	delete(body, BodyRevisions)
-
-	newDoc.UpdateBody(body)
+	generation, _ := ParseRevID(newDoc.RevID)
```

Comment:
See comments below about moving validation into `Put` instead of `PutWithBody`

```suggestion
	generation, _ := ParseRevID(newDoc.RevID)
	if generation < 0 {
		return "", nil, base.HTTPErrorf(http.StatusBadRequest, "Invalid revision ID")
	} 
```